src/Features/Core/Portable/NavigateTo/AbstractNavigateToSearchService.cs

```diff
@@ -62,4 +67,91 @@ private static PooledDisposer<PooledHashSet<T>> GetPooledHashSet<T>(ImmutableArr
+        // Concurrently, the reading task will grab items when available, and send them over to the host.
+        var channel = Channel.CreateUnbounded<RoslynNavigateToItem>(s_channelOptions);
+
+        await Task.WhenAll(
+            FindAllItemsAndWriteToChannelAsync(),
+            ReadItemsFromChannelAndReportToCallbackAsync()).ConfigureAwait(false);
+
+        async Task ReadItemsFromChannelAndReportToCallbackAsync()
+        {
+            await Task.Yield().ConfigureAwait(false);
+            using var _ = ArrayBuilder<RoslynNavigateToItem>.GetInstance(out var items);
+
+            while (await channel.Reader.WaitToReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                // Grab as many items as we can from the channel at once and report in a batch.
+                while (channel.Reader.TryRead(out var item))
+                    items.Add(item);
+
+                await onItemsFound(items.ToImmutableAndClear()).ConfigureAwait(false);
+            }
+        }
+
+        async Task FindAllItemsAndWriteToChannelAsync()
+        {
+            Exception? exception = null;
+            try
+            {
+                await Task.Yield().ConfigureAwait(false);
+                await PerformSearchAsync(item => channel.Writer.TryWrite(item)).ConfigureAwait(false);
+            }
+            catch (Exception ex) when ((exception = ex) == null)
+            {
+                throw ExceptionUtilities.Unreachable();
+            }
+            finally
+            {
+                // No matter what path we take (exceptional or non-exceptional), always complete the channel so the
+                // writing task knows it's done.
+                channel.Writer.TryComplete(exception);
+            }
+        }
+
+        Task PerformSearchAsync(Action<RoslynNavigateToItem> onItemFound)
+            => ParallelForEachAsync(
+                items, cancellationToken, (item, cancellationToken) => callback(item, onItemFound, cancellationToken));
+    }
+
+#pragma warning disable CA1068 // CancellationToken parameters must come last
+    private static async Task ParallelForEachAsync<TSource>(
+#pragma warning restore CA1068 // CancellationToken parameters must come last
+        IEnumerable<TSource> source,
+        CancellationToken cancellationToken,
+        Func<TSource, CancellationToken, ValueTask> body)
+    {
+        if (cancellationToken.IsCancellationRequested)
+            return;
+
+#if NET
+        await Parallel.ForEachAsync(source, cancellationToken, body).ConfigureAwait(false);
+#else
+        using var _ = ArrayBuilder<Task>.GetInstance(out var tasks);
+
+        foreach (var item in source)
+        {
+            tasks.Add(Task.Run(async () =>
+            {
+                await body(item, cancellationToken).ConfigureAwait(false);
+            }, cancellationToken));
+        }
+
+        await Task.WhenAll(tasks).ConfigureAwait(false);
```

Comment:
poor man's solution.  I consider copying the runtime code and forking into roslyn.  but it was complex enough i didn't want us to main that.  And that code should not run here as this is the searching part, which should run in OOP on VS and on netcore for vscode anyways.   (And it's the same logic as before, so no worse for any other client that was previously using hte old codepaths).